### PR TITLE
[PM-15084] Push global notification creation to affected clients

### PR DESCRIPTION
**Pull request by mzieniukbw**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-15084

This is Part 4.
See Part 3: #5057 

## 📔 Objective

When a global notification is created via the service layer, distributes a “notification created” or "notification updated" push type event to related clients of the user.
Global notifications are sent to servers of the same installation id. All clients connected to any of the server instance of that installation id would receive it.
This is useful for notifying all clients of an installation about upcoming maintenance.
This works both for Self-Hosted, but also for Cloud (bitwarden.com), assuming an installation id is set.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: test/Common/AutoFixture/Attributes/RepeatingPatternBitAutoDataAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+﻿using System.Reflection;
+
+namespace Bit.Test.Common.AutoFixture.Attributes;
+
+public class RepeatingPatternBitAutoDataAttribute : BitAutoDataAttribute
```
Comment (Thomas-Avery) [inline test/Common/AutoFixture/Attributes/RepeatingPatternBitAutoDataAttribute.cs:5]: 💭 
I really like the idea of what we're trying to solve here. It's annoying to type out and easy to miss some of those inline data sets. 

I think it would have been easier to have this in a separate PR, since it's introducing a new testing tool/attribute. 



Xunit has a few different solutions with their idea of sharing test data `ClassData` `MemberData` and I know there are a few libraries that similarly do what you have proposed.


For this specifically, I think combining xUnit's idea of `ClassData` with `BitAutoData` is an alternative to consider here. 

Something like (this is rough code could be cleaned up)

```
  public class ClassBitAutoDataAttribute : BitAutoDataAttribute
    {
        public ClassBitAutoDataAttribute(Type @class)
        {
            Class = @class;
        }

        public Type Class { get; private set; }

        public override IEnumerable<object[]> GetData(MethodInfo testMethod)
        {
            IEnumerable<object[]> data = Activator.CreateInstance(Class) as IEnumerable<object[]>;
            if (data == null)
            {
                throw new ArgumentException(
                    string.Format(
                        CultureInfo.CurrentCulture,
                        "{0} must implement IEnumerable<object[]> to be used as ClassData for the test method named '{1}' on {2}",
                        Class.FullName,
                        testMethod.Name,
                        testMethod.DeclaringType.FullName
                    )
                );
            }

            foreach (var classData in data)
            {
                var bitData = base.GetData(testMethod).First();
                for (var i = 0; i < classData.Length; i++)
                {
                    bitData[i] = classData[i];
                }

                yield return bitData;
            }
        }
    }
```


Then we could have one class to share data across tests

```
   public class SendAsyncTestData : TheoryData<bool, bool, bool>
    {
        public SendAsyncTestData()
        {
            Add(false, false, false);
            Add(true, false, false);
            Add(false, true, false);
            Add(true, true, false);
            Add(false, false, true);
            Add(true, false, true);
            Add(false, true, true);
        }
    };
```

Usage in an individual test.

```
    [Theory]
    [ClassBitAutoDataAttribute(typeof(SendAsyncTestData))]
    public async Task SendAsync_UserIdSet_SendPayloadToUserAsync(bool haveIdentifier, bool haveDeviceId,
        bool haveOrganizationId, SutProvider<PushController> sutProvider, Guid installationId, Guid userId,
        Guid identifier, Guid deviceId)
    {
```

This would give us more flexibility and allow for generating (looping through combinations) the test data within the test data class if required.

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
Looking good, a few things to take a look at.

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs [hunk 2]
```diff
@@ -192,26 +205,36 @@ public async Task PushNotificationAsync(Notification notification)
...
             OrganizationId = notification.OrganizationId,
+            InstallationId = installationId,
             Title = notification.Title,
             Body = notification.Body,
             CreationDate = notification.CreationDate,
             RevisionDate = notification.RevisionDate
         };
 
-        if (notification.UserId.HasValue)
+        if (notification.Global && installationId.HasValue)
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs:221]: Chose to log warning in constructor, since changing installation it to be required and throwing exceptions is large change. I will add it as an option to https://bitwarden.atlassian.net/browse/PM-15513
Fixed in 186a09bb

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs [hunk 3]
```diff
@@ -228,15 +252,20 @@ public async Task PushNotificationStatusAsync(Notification notification, Notific
             DeletedDate = notificationStatus.DeletedDate
         };
 
-        if (notification.UserId.HasValue)
+        if (notification.Global && installationId.HasValue)
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs:261]: Fixed in 186a09bb

File: test/Common/AutoFixture/Attributes/RepeatingPatternBitAutoDataAttribute.cs (hunk 1 above)
Comment (mzieniukbw) [reply test/Common/AutoFixture/Attributes/RepeatingPatternBitAutoDataAttribute.cs:5]: Separated to #5167

I don't like the `SendAsyncTestData` suggestion, since it obscures the data from the test, so it's harder to know what we are actually testing. And i know we will have `SendAsyncTestData2Booleans`, `SendAsyncTestData3Booleans` and other more complex data's that will be hard to describe.

Comment (Thomas-Avery) [reply test/Common/AutoFixture/Attributes/RepeatingPatternBitAutoDataAttribute.cs:5]: Agreed on the naming and repeat of class test data, but that is just a general concern of using that XUnit feature.


If obscurity is a concern, I don't see how either option is an improvement. Tracking down what data was used in the test that is intermittently failing is a problem we face with using BitAutoData in the current state.

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs:221]: Not exactly what I was asking about. 

```
    if (notification.Global && installationId.HasValue)
        {
            await SendPayloadToInstallationAsync(installationId.Value, PushType.Notification, message, true,
                notification.ClientType);
        }
        else if (notification.UserId.HasValue)
        {
            await SendPayloadToUserAsync(notification.UserId.Value, PushType.Notification, message, true,
                notification.ClientType);
        }
        else if (notification.OrganizationId.HasValue)
        {
            await SendPayloadToOrganizationAsync(notification.OrganizationId.Value, PushType.Notification, message,
                true, notification.ClientType);
        }
        else 
        {
           /// Currently siliently does nothing should we add warning logging here?
        }
    }
```

This would specifically be a miss configured notification which shouldn't happen, but my preference is to handle the unexpected case explicitly with either an exception or a log instead of ignoring it.

**Comment by mzieniukbw** [comment]
Must have messed up the merge. Needs to rebase

File: util/Migrator/DbScripts/2024-12-04_00_AddActiveDeviceWithPendingAuth.sql [hunk 4]
```diff
@@ -0,0 +1,27 @@
+CREATE OR ALTER PROCEDURE [dbo].[Device_ReadActiveWithPendingAuthRequestsByUserId]
+    @UserId UNIQUEIDENTIFIER,
```
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-12-04_00_AddActiveDeviceWithPendingAuth.sql:2]: @trmartin4 will this have same issue as https://github.com/bitwarden/server/blob/main/src/Sql/Auth/dbo/Stored%20Procedures/Device_ReadActiveWithPendingAuthRequestsByUserId.sql

Comment (trmartin4) [reply util/Migrator/DbScripts/2024-12-04_00_AddActiveDeviceWithPendingAuth.sql:2]: @rkac-bw I believe this is the same stored procedure, just reflected in this branch (which is part of a PR stack, so not targeting `main`).

I created a second migration for the changes to update the SP, and that is in `main` and should be reflected when `main` is merged up into this branch.

Let me know if you feel like I've missed something and need to address this separately.

**Comment by withinfocus** [comment]
Still needing rebase right?

**Comment by mzieniukbw** [comment]
I had to rebase and force push on top of `main`, since i could not get `git merge` working. Nothing have changed, since last time this PR got approved. 
I have also self-reviewed the changed to make sure everything is as intended.

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs]: Added a log warning on those edge cases. Fixed in 7e6501c3

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs]: Added a log warning on those edge cases. Fixed in 7e6501c3

File: test/Core.Test/Platform/Push/Services/AzureQueuePushNotificationServiceTests.cs [hunk 5]
```diff
@@ -14,7 +16,7 @@
 using NSubstitute;
 using Xunit;
 
-namespace Bit.Core.Platform.Push.Internal.Test;
+namespace Bit.Core.Test.Platform.Push.Internal;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/AzureQueuePushNotificationServiceTests.cs:19]: Entire `Platform` namespace is placed in wrong place, did not want to change it. But might as well refactor it here.

[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply test/Core.Test/Platform/Push/Services/AzureQueuePushNotificationServiceTests.cs:19]: I'm fine with leaving that out of scope. These tests ones can be altered without effecting those.

File: src/Core/Platform/Push/Services/AzureQueuePushNotificationService.cs [hunk 6]
```diff
@@ -7,25 +7,36 @@
...
-        IHttpContextAccessor httpContextAccessor)
+        IHttpContextAccessor httpContextAccessor,
+        IGlobalSettings globalSettings,
+        ILogger<AzureQueuePushNotificationService> logger)
     {
         _queueClient = queueClient;
         _httpContextAccessor = httpContextAccessor;
+        _globalSettings = globalSettings;
+
+        if (globalSettings.Installation.Id == default)
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/Platform/Push/Services/AzureQueuePushNotificationService.cs]: Fixed in ddcf7d89

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs [hunk 7]
```diff
@@ -185,6 +199,10 @@ public async Task PushAuthRequestResponseAsync(AuthRequest authRequest)
 
     public async Task PushNotificationAsync(Notification notification)
     {
+        Guid? installationId = notification.Global && _globalSettings.Installation.Id != default
+            ? _globalSettings.Installation.Id
+            : null;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs]: Fixed in ddcf7d89

File: src/Core/NotificationHub/NotificationHubPushNotificationService.cs [hunk 8]
```diff
@@ -18,24 +19,37 @@
...
+        ILogger<NotificationHubPushNotificationService> logger,
+        IGlobalSettings globalSettings)
     {
         _installationDeviceRepository = installationDeviceRepository;
         _httpContextAccessor = httpContextAccessor;
         _notificationHubPool = notificationHubPool;
         _logger = logger;
+        _globalSettings = globalSettings;
+
+        if (globalSettings.Installation.Id == default)
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/NotificationHub/NotificationHubPushNotificationService.cs]: Fixed in ddcf7d89

File: src/SharedWeb/Utilities/ServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -280,9 +280,13 @@ public static void AddDefaultServices(this IServiceCollection services, GlobalSe
         services.AddSingleton<IPushNotificationService, MultiServicePushNotificationService>();
         if (globalSettings.SelfHosted)
         {
+            if (globalSettings.Installation.Id == default)
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/SharedWeb/Utilities/ServiceCollectionExtensions.cs]: Fixed in ddcf7d89

File: test/Core.Test/Platform/Push/Services/AzureQueuePushNotificationServiceTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/AzureQueuePushNotificationServiceTests.cs]: Fixed in 711eb157

File: test/Core.Test/Platform/Push/Services/MultiServicePushNotificationServiceTests.cs [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
+using Bit.Core.Platform.Push;
+using Bit.Core.Platform.Push.Internal;
 using Bit.Core.Test.NotificationCenter.AutoFixture;
 using Bit.Test.Common.AutoFixture;
 using Bit.Test.Common.AutoFixture.Attributes;
 using NSubstitute;
 using Xunit;
 
-namespace Bit.Core.Platform.Push.Internal.Test;
+namespace Bit.Core.Test.Platform.Push.Internal;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/MultiServicePushNotificationServiceTests.cs]: Fixed in 711eb157

File: test/Core.Test/Platform/Push/Services/NotificationsApiPushNotificationServiceTests.cs [hunk 11]
```diff
@@ -1,10 +1,11 @@
-﻿using Bit.Core.Settings;
+﻿using Bit.Core.Platform.Push;
+using Bit.Core.Settings;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.Logging;
 using NSubstitute;
 using Xunit;
 
-namespace Bit.Core.Platform.Push.Internal.Test;
+namespace Bit.Core.Test.Platform.Push.Internal;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/NotificationsApiPushNotificationServiceTests.cs]: Fixed in 711eb157

File: test/Api.Test/Platform/Push/Controllers/PushControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+using Bit.Core.Exceptions;
+using Bit.Core.Models.Api;
+using Bit.Core.Platform.Push;
+using Bit.Core.Settings;
+using Bit.Test.Common.AutoFixture;
+using Bit.Test.Common.AutoFixture.Attributes;
+using NSubstitute;
+using Xunit;
+
+namespace Bit.Api.Test.Platform.Push;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Api.Test/Platform/Push/Controllers/PushControllerTests.cs]: Fixed in 711eb157

File: test/Core.Test/Platform/Push/Services/RelayPushRegistrationServiceTests.cs [hunk 13]
```diff
@@ -1,9 +1,10 @@
-﻿using Bit.Core.Settings;
+﻿using Bit.Core.Platform.Push.Internal;
+using Bit.Core.Settings;
 using Microsoft.Extensions.Logging;
 using NSubstitute;
 using Xunit;
 
-namespace Bit.Core.Platform.Push.Internal.Test;
+namespace Bit.Core.Test.Platform.Push.Internal;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/RelayPushRegistrationServiceTests.cs]: Fixed in 711eb157

File: test/Core.Test/Platform/Push/Services/RelayPushNotificationServiceTests.cs [hunk 14]
```diff
@@ -1,11 +1,12 @@
...
+﻿using Bit.Core.Platform.Push.Internal;
+using Bit.Core.Repositories;
 using Bit.Core.Settings;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.Logging;
 using NSubstitute;
 using Xunit;
 
-namespace Bit.Core.Platform.Push.Internal.Test;
+namespace Bit.Core.Test.Platform.Push.Internal;
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply test/Core.Test/Platform/Push/Services/RelayPushNotificationServiceTests.cs]: Fixed in 711eb157

**Review by Thomas-Avery — APPROVED** [review]

**Comment by mzieniukbw** [comment]
Need to resolve some conflicts, something went wrong during merge.

**Review by justindbaur — COMMENTED** [review]
Just one question.

File: src/Core/Models/Api/Request/PushRegistrationRequestModel.cs [hunk 15]
```diff
@@ -5,15 +5,11 @@ namespace Bit.Core.Models.Api;
...
-    public DeviceType Type { get; set; }
-    [Required]
-    public string Identifier { get; set; }
+    [Required] public string DeviceId { get; set; }
+    [Required] public string PushToken { get; set; }
+    [Required] public string UserId { get; set; }
+    [Required] public DeviceType Type { get; set; }
+    [Required] public string Identifier { get; set; }
     public IEnumerable<string> OrganizationIds { get; set; }
+    public string InstallationId { get; set; }
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/Models/Api/Request/PushRegistrationRequestModel.cs:14]: No idea. This is used in Controller as body, other fields are also string. Probably legacy reasons ? Do you want me to change it ? I just wonder if this will work as is.

[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/Models/Api/Request/PushRegistrationRequestModel.cs]: Fixed in a9bf2e19

[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Models/Api/Request/PushRegistrationRequestModel.cs]: So sorry, does a nullable `Guid` fit even better here?

**Review by Thomas-Avery — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

File: src/Core/Models/Api/Request/PushRegistrationRequestModel.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply src/Core/Models/Api/Request/PushRegistrationRequestModel.cs]: I will address it in separate PR, since i have more refactoring like changes coming.
